pdftk.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH PDFTK 1 "December 7, 2020"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.ad l
.SH NAME
pdftk \- A handy tool for manipulating PDF
.SH SYNOPSIS
\fBpdftk\fR \fI<input PDF files | \- | PROMPT>\fR
.br
     [ \fBinput_pw\fR \fI<input PDF owner passwords | PROMPT>\fR ]
.br
     [ \fI<operation>\fR \fI<operation arguments>\fR ]
.br
     [ \fBoutput\fR \fI<output filename | \- | PROMPT>\fR ]
.br
     [ \fBencrypt_40bit\fR | \fBencrypt_128bit\fR | \fBencrypt_aes128\fR ]
.br
     [ \fBallow\fR \fI<permissions>\fR ]
.br
     [ \fBowner_pw\fR \fI<owner password | PROMPT>\fR ] 
.br
     [ \fBuser_pw\fR \fI<user password | PROMPT>\fR ]
.br
     [ \fBflatten\fR ] [ \fBneed_appearances\fR ] 
.br
     [ \fBcompress\fR | \fBuncompress\fR ]
.br
     [ \fBkeep_first_id\fR | \fBkeep_final_id\fR ] [ \fBdrop_xfa\fR ] [ \fBdrop_xmp\fR ]
.br
     [ \fBreplacement_font\fR \fI<font name>\fR ]
.br
     [ \fBverbose\fR ] [ \fBdont_ask\fR | \fBdo_ask\fR ]
.br
Where:
.br
     \fI<operation>\fR may be empty, or:
.br
     [ \fBcat\fR | \fBshuffle\fR | \fBburst\fR | \fBrotate\fR |
.br
       \fBgenerate_fdf\fR | \fBfill_form\fR |
.br
       \fBbackground\fR | \fBmultibackground\fR |
.br
       \fBstamp\fR | \fBmultistamp\fR |
.br
       \fBdump_data\fR | \fBdump_data_utf8\fR |
.br
       \fBdump_data_fields\fR | \fBdump_data_fields_utf8\fR |
.br
       \fBdump_data_annots\fR |
.br
       \fBupdate_info\fR | \fBupdate_info_utf8\fR |
.br
       \fBattach_files\fR | \fBunpack_files\fR ]
.br

For Complete Help: \fBpdftk \-\-help\fR
.br
.SH DESCRIPTION
If PDF is electronic paper, then pdftk is an electronic staple-remover, hole-punch, binder, secret-decoder-ring, and X-Ray-glasses.  Pdftk is a simple tool for doing everyday things with PDF documents.  Use it to:
.sp
.br
* Merge PDF Documents or Collate PDF Page Scans
.br
* Split PDF Pages into a New Document
.br
* Rotate PDF Documents or Pages
.br
* Decrypt Input as Necessary (Password Required)
.br
* Encrypt Output as Desired
.br
* Fill PDF Forms with X/FDF Data and/or Flatten Forms
.br
* Generate FDF Data Stencils from PDF Forms
.br
* Apply a Background Watermark or a Foreground Stamp
.br
* Report PDF Metrics, Bookmarks and Metadata
.br
* Add/Update PDF Metrics, Bookmarks or Metadata
.br
* Attach Files to PDF Pages or the PDF Document
.br
* Unpack PDF Attachments
.br
* Burst a PDF Document into Single Pages
.br
* Uncompress and Re-Compress Page Streams
.br
* Repair Corrupted PDF (Where Possible)
.SH OPTIONS
A summary of options is included below.
.TP
\fB\-\-help\fR, \fB\-h\fR
Show this summary of options.
.TP
.B <input PDF files | \- | PROMPT>
A list of the input PDF files. If you plan to combine these PDFs (without
using handles) then list files in the order you want them combined.  Use \fB\-\fR 
to pass a single PDF into pdftk via stdin.
Input files can be associated with handles, where a
handle is one or more upper-case letters:

\fI<input PDF handle>\fR\fB=\fR\fI<input PDF filename>\fR

Handles are often omitted.  They are useful when specifying PDF passwords or page ranges, later.

For example: A=input1.pdf QT=input2.pdf M=input3.pdf
.TP
.B [input_pw <input PDF owner passwords | PROMPT>]
Input PDF owner passwords, if necessary, are associated with files
by using their handles:

\fI<input PDF handle>\fR\fB=\fR\fI<input PDF file owner password>\fR

If handles are not given, then passwords are associated with input
files by order.

Most pdftk features require that encrypted 
input PDF are accompanied by the ~owner~ password. If the input PDF
has no owner password, then the user password must be given, instead.
If the input PDF has no passwords, then no password should be given.

When running in \fBdo_ask\fR mode, pdftk will prompt you for a password
if the supplied password is incorrect or none was given.
.TP
.B [<operation> <operation arguments>]
Available operations are: \fBcat\fR, \fBshuffle\fR, \fBburst\fR, \fBrotate\fR,
\fBgenerate_fdf\fR, \fBfill_form\fR, \fBbackground\fR, \fBmultibackground\fR, 
\fBstamp\fR, \fBmultistamp\fR, \fBdump_data\fR, \fBdump_data_utf8\fR, 
\fBdump_data_fields\fR, \fBdump_data_fields_utf8\fR, \fBdump_data_annots\fR, \fBupdate_info\fR, 
\fBupdate_info_utf8\fR, \fBattach_files\fR, \fBunpack_files\fR. Some operations
takes additional arguments, described below.

If this optional argument is omitted, then pdftk runs in 'filter' mode.
Filter mode takes only one PDF input and creates a new PDF after
applying all of the output options, like encryption and compression.
.RS 3
.TP
.B cat [<page ranges>]
Assembles (catenates) pages from input PDFs to create a new PDF. Use \fBcat\fR to merge PDF pages or to split PDF pages from documents. You can also use it to rotate PDF pages. Page order in the new PDF is specified by the order of the given page ranges. Page ranges are described like this:

\fI<input PDF handle>\fR[\fI<begin page number>\fR[\fB\-\fR\fI<end page number>\fR[\fI<qualifier>\fR]]][\fI<page rotation>\fR]

Where the handle identifies one of the input PDF files, and
the beginning and ending page numbers are one-based references
to pages in the PDF file.
The qualifier can be \fBeven\fR, \fBodd\fR, or \fB~\fR, and the page rotation can be \fBnorth\fR, \fBsouth\fR, \fBeast\fR, \fBwest\fR, \fBleft\fR, \fBright\fR, or \fBdown\fR.

If a PDF handle is given but no pages are specified, then the entire PDF is used. If no pages are specified for any of the input PDFs, then the input PDFs' bookmarks are also merged and included in the output.

If the handle is omitted from the page range, then the pages are taken from the first input PDF.

The \fBeven\fR qualifier causes pdftk to use only the even-numbered PDF pages, so \fB1\-6even\fR yields pages 2, 4 and 6 in that order.  \fB6\-1even\fR yields pages 6, 4 and 2 in that order.

The \fBodd\fR qualifier works similarly to the \fBeven\fR.

Pages can be subtracted from a page range using the \fB~\fR qualifier
followed by a page range. For instance, \fB1\-20~5\-6\fR and
\fB1\-20~5~6\fR are equivalent to \fB1\-4 7\-20\fR, and \fB~5\fR yields
all pages except page 5. Depending on your shell, you may need to
quote this argument because of the ~ at the beginning.

The page rotation setting can cause pdftk to rotate pages and documents.  Each option sets the page rotation as follows (in degrees): \fBnorth\fR: 0, \fBeast\fR: 90, \fBsouth\fR: 180, \fBwest\fR: 270, \fBleft\fR: \-90, \fBright\fR: +90, \fBdown\fR: +180. \fBleft\fR, \fBright\fR, and \fBdown\fR make relative adjustments to a page's rotation.

If no arguments are passed to cat, then pdftk combines all input PDFs in the
order they were given to create the output.

.B NOTES:
.br
* \fI<end page number>\fR may be less than \fI<begin page number>\fR.
.br
* The keyword \fBend\fR may be used to reference the final page of a document instead of a page number.
.br
* Reference a single page by omitting the ending page number.
.br
* The handle may be used alone to represent the entire PDF document, e.g., B1\-end is the same as B.
.br
* You can reference page numbers in reverse order by prefixing them with the letter \fBr\fR. For example, page r1 is the last page of the document, r2 is the next-to-last page of the document, and rend is the first page of the document. You can use this prefix in ranges, too, for example r3\-r1 is the last three pages of a PDF.

.B Page Range Examples without Handles:
.br
\fB1\\-endeast\fR \(en rotate entire document 90 degrees
.br
\fB5 11 20\fR \(en take single pages from input PDF
.br
\fB5\-25oddwest\fR \(en take odd pages in range, rotate 90 degrees
.br
\fB6\-1\fR \(en reverse pages in range from input PDF

.B Page Range Examples Using Handles:
.br
Say \fBA=in1.pdf B=in2.pdf\fR, then:
.br
\fBA1\-21\fR \(en take range from in1.pdf
.br
\fBBend\-1odd\fR \(en take all odd pages from in2.pdf in reverse order
.br
\fBA72\fR \(en take a single page from in1.pdf
.br
\fBA1\-21 Beven A72\fR \(en assemble pages from both in1.pdf and in2.pdf
.br
\fBAwest\fR \(en rotate entire in1.pdf document 90 degrees
.br
\fBB\fR \(en use all of in2.pdf
.br
\fBA2\-30evenleft\fR \(en take the even pages from the range, remove 90 degrees from each page's rotation
.br
\fBA A\fR \(en catenate in1.pdf with in1.pdf
.br
\fBAevenwest Aoddeast\fR \(en apply rotations to even pages, odd pages from in1.pdf
.br
\fBAwest Bwest Bdown\fR \(en catenate rotated documents
.TP
.B shuffle [<page ranges>]
Collates pages from input PDFs to create a new PDF.  Works like the \fBcat\fR operation except that it takes one page at a time from each page range to assemble the output PDF.  If one range runs out of pages, it continues with the remaining ranges.  Ranges can use all of the features described above for \fBcat\fR, like reverse page ranges, multiple ranges from a single PDF, and page rotation.  This feature was designed to help collate PDF pages after scanning paper documents.
.TP
.B burst
Splits a single input PDF document into individual pages. Also creates a
report named \fBdoc_data.txt\fR which is the same as the output from \fBdump_data\fR.
The \fBoutput\fR section can contain a printf-styled format string to name these pages.
For example, if you want pages named page_01.pdf, page_02.pdf, etc., pass
\fBoutput page_%02d.pdf\fR to pdftk. If the pattern is omitted, then a default pattern
\fpg_%04d.pdf\fR is appended and produces pages named pg_0001.pdf, pg_0002.pdf, etc.
Encryption can be applied to the output by appending output options such as \fBowner_pw\fR, e.g.:

pdftk in.pdf burst owner_pw foopass
.TP
.B rotate [<page ranges>]
Takes a single input PDF and rotates just the specified pages.  All other pages remain unchanged.  The page order remains unchanged.  Specify the pages to rotate using the same notation as you would with \fBcat\fR, except you omit the pages that you aren't rotating:

[\fI<begin page number>\fR[\fB\-\fR\fI<end page number>\fR[\fI<qualifier>\fR]]][\fI<page rotation>\fR]

The qualifier can be \fBeven\fR or \fBodd\fR, and the page rotation can be \fBnorth\fR, \fBsouth\fR, \fBeast\fR, \fBwest\fR, \fBleft\fR, \fBright\fR, or \fBdown\fR.

Each option sets the page rotation as follows (in degrees): \fBnorth\fR: 0, \fBeast\fR: 90, \fBsouth\fR: 180, \fBwest\fR: 270, \fBleft\fR: \-90, \fBright\fR: +90, \fBdown\fR: +180. \fBleft\fR, \fBright\fR, and \fBdown\fR make relative adjustments to a page's rotation.

The given order of the pages doesn't change the page order in the output.
.TP
.B generate_fdf
Reads a single input PDF file and generates an FDF file suitable for \fBfill_form\fR
out of it to the given output
filename or (if no output is given) to stdout.  Does not create a new PDF.
.TP
.B fill_form <FDF data filename | XFDF data filename | \- | PROMPT>
Fills the single input PDF's form fields with the data from an FDF file, XFDF file or stdin. Enter the data filename
after \fBfill_form\fR, or use \fB\-\fR to pass the data via stdin, like so:

pdftk form.pdf fill_form data.fdf output form.filled.pdf

If the input FDF file includes Rich Text formatted data in addition to plain text, then the
Rich Text data is packed into the form fields \fIas well as\fR the plain text.  Pdftk also sets a flag
that cues Reader/Acrobat to generate new field appearances based on the Rich Text data.  So
when the user opens the PDF, the viewer will create the Rich Text appearance on the spot.  If the
user's PDF viewer does not support Rich Text, then the user will see the plain text data instead.
If you flatten this form before Acrobat has a chance to create (and save) new field appearances,
then the plain text field data is what you'll see.

Also see the \fBflatten\fR, \fBneed_appearances\fR, and \fBreplacement_font\fR options.
.TP
.B background <background PDF filename | \- | PROMPT>
Applies a PDF watermark to the background of a single input PDF.  Pass the background PDF's
filename after \fBbackground\fR like so:

pdftk in.pdf background back.pdf output out.pdf

Pdftk uses only the first page from the background PDF and applies it to every page of the
input PDF.  This page is scaled and rotated as needed to fit the input page.  You can use \fB\-\fR
to pass a background PDF into pdftk via stdin.

If the input PDF does not have a transparent background (such as a PDF created from page scans) then the resulting background won't be visible \(en use the \fBstamp\fR operation instead.
.TP
.B multibackground <background PDF filename | \- | PROMPT>
Same as the \fBbackground\fR operation, but applies each page of the background PDF to the corresponding page of the input PDF.  If the input PDF has more pages than the stamp PDF, then the final stamp page is repeated across these remaining pages in the input PDF.
.TP
.B stamp <stamp PDF filename | \- | PROMPT>
This behaves just like the \fBbackground\fR operation except it overlays the stamp PDF page \fIon top\fR of the input PDF document's pages.  This works best if the stamp PDF page has a transparent background.
.TP
.B multistamp <stamp PDF filename | \- | PROMPT>
Same as the \fBstamp\fR operation, but applies each page of the background PDF to the corresponding page of the input PDF.  If the input PDF has more pages than the stamp PDF, then the final stamp page is repeated across these remaining pages in the input PDF.
.TP
.B dump_data
Reads a single input PDF file and reports its metadata, bookmarks (a/k/a outlines), page metrics (media, rotation and labels), data embedded by STAMPtk (see STAMPtk's \fBembed\fR option) and other data to the given output filename or (if no output is given) to stdout.  Non-ASCII characters are encoded as XML numerical entities.  Does not create a new PDF.
.TP
.B dump_data_utf8
Same as \fBdump_data\fR except that the output is encoded as UTF-8.
.TP
.B dump_data_fields
Reads a single input PDF file and reports form field statistics to the given output
filename or (if no output is given) to stdout. Non-ASCII characters are encoded
as XML numerical entities. Does not create a new PDF.
.TP
.B dump_data_fields_utf8
Same as \fBdump_data_fields\fR except that the output is encoded as UTF-8.
.TP
.B dump_data_annots
\fBThis operation currently reports only link annotations.\fR
Reads a single input PDF file and reports annotation information to the given output
filename or (if no output is given) to stdout. Non-ASCII characters are encoded
as XML numerical entities. Does not create a new PDF.
.TP
.B update_info <info data filename | \- | PROMPT>
Changes the bookmarks, page labels, page sizes, page rotations, and
metadata in a single PDF's Info dictionary to match the input data
file. The input data file uses the same syntax as the output from
\fBdump_data\fR. Non-ASCII characters should be encoded as XML
numerical entities.

This operation does not change the metadata stored
in the PDF's XMP stream, if it has one. (For this reason you should include
a \fBModDate\fR entry in your updated info with a current date/timestamp, format:
\fBD:YYYYMMDDHHmmSS\fR, e.g. D:201307241346 \(en omitted data after YYYY revert 
to default values.)

For example:

pdftk in.pdf update_info in.info output out.pdf
.TP
.B update_info_utf8 <info data filename | \- | PROMPT>
Same as \fBupdate_info\fR except that the input is encoded as UTF-8.
.TP
.B attach_files <attachment filenames | PROMPT> [to_page <page number | PROMPT> | relation <relationship>]
Packs arbitrary files into a PDF using PDF's file attachment
features. More than one attachment may be listed after
\fBattach_files\fR. Attachments are added at the document level unless
the optional \fBto_page\fR option is given, in which case the files
are attached to the given page number (the first page is 1, the final
page is \fBend\fR). Attachments at the document level may be tagged
with a relationship among \fBSource\fR, \fBData\fR, \fBAlternative\fR,
\fBSupplement\fR, and \fBUnspecified\fR (default).

For example:

pdftk in.pdf attach_files table1.html table2.html to_page 6 output out.pdf

pdftk in.pdf attach_files in.tex relation Source output out.pdf
.TP
.B unpack_files
Copies all of the attachments from the input PDF into the current folder or to
an output directory given after \fBoutput\fR. For example:

pdftk report.pdf unpack_files output ~/atts/

or, interactively:

pdftk report.pdf unpack_files output PROMPT
.RE
.TP
.B [output <output filename | \- | PROMPT>]
The output PDF filename may not be set to the name of an input filename. Use
\fB\-\fR to output to stdout.
When using the \fBdump_data\fR operation, use \fBoutput\fR to set the name of the
output data file. When using the \fBunpack_files\fR operation, use \fBoutput\fR to set
the name of an output directory.  When using the \fBburst\fR operation, you can use \fBoutput\fR
to control the resulting PDF page filenames (described above).
.TP
.B [encrypt_40bit | encrypt_128bit | encrypt_aes128]
If an output PDF user or owner password is given, the output PDF
encryption algorithm defaults to AES-128. The weaker RC4 40-bit and
RC4 128-bit algorithms can be chosen by specifying encrypt_40bit or
encrypt_128bit (discouraged).
.TP
.B [allow <permissions>]
Permissions are applied to the output PDF only if an encryption strength
is specified or an owner or user password is given.  If permissions are
not specified, they default to 'none,' which means all of the following
features are disabled.

The \fBpermissions\fR section may include one or more of the following 
features:
.RS
.TP 
.B Printing
Top Quality Printing
.TP
.B DegradedPrinting
Lower Quality Printing
.TP 
.B ModifyContents
Also allows Assembly
.TP 
.B Assembly
.TP
.B CopyContents
Also allows ScreenReaders
.TP
.B ScreenReaders
.TP
.B ModifyAnnotations
Also allows FillIn
.TP
.B FillIn
.TP
.B AllFeatures
Allows the user to perform all of the above, and top quality printing.
.RE
.TP
.B [owner_pw <owner password | PROMPT>]
.TP
.B [user_pw <user password | PROMPT>]
If an encryption strength is given but no passwords are supplied, then
the owner and user passwords remain empty, which means that the resulting
PDF may be opened and its security parameters altered by anybody.
.TP
.B [compress | uncompress]
These are only useful when you want to edit PDF code in a text editor like vim or emacs.
Remove PDF page stream compression by
applying the \fBuncompress\fR filter. Use the \fBcompress\fR filter to restore compression.
.TP
.B [flatten]
Use this option to merge an input PDF's interactive form fields (and their data) with
the PDF's pages. Only one input PDF may be given. Sometimes used with the \fBfill_form\fR operation.
.TP
.B [need_appearances]
Sets a flag that cues Reader/Acrobat to generate new field appearances based on the form field values.  Use this when filling a form with non-ASCII text to ensure the best presentation in Adobe Reader or Acrobat.  It won't work when combined with the \fBflatten\fR option.
.TP
.B [replacement_font <font name>]
Use the specified font to display text in form fields. This option is useful when filling a form with non-ASCII text that is not supported by the fonts included in the input PDF. \fBfont name\fR may be either the file name or the family name of a font, but using a file name is more reliable. Currently only TrueType fonts with Unicode text are supported.
.TP
.B [keep_first_id | keep_final_id]
When combining pages from multiple PDFs, use one of these options to copy the document ID from either the first or final input document into the new output PDF. Otherwise pdftk creates a new document ID for the output PDF. When no operation is given, pdftk always uses the ID from the (single) input PDF.
.TP
.B [drop_xfa]
If your input PDF is a form created using Acrobat 7 or Adobe Designer, then it probably has XFA data.  Filling such a form using pdftk yields a PDF with data that fails to display in Acrobat 7 (and 6?).  The workaround solution is to remove the form's XFA data, either before you fill the form using pdftk or at the time you fill the form. Using this option causes pdftk to omit the XFA data from the output PDF form.

This option is only useful when running pdftk on a single input PDF.  When assembling a PDF from multiple inputs using pdftk, any XFA data in the input is automatically omitted.
.TP
.B [drop_xmp]
Many PDFs store document metadata using both an Info dictionary (old school) and an XMP stream (new school).  Pdftk's \fBupdate_info\fR operation can update the Info dictionary, but not the XMP stream.  The proper remedy for this is to include a \%\fBModDate\fR entry in your updated info with a current date/timestamp. The date/timestamp format is: \fBD:YYYYMMDDHHmmSS\fR, e.g. D:201307241346 \(en omitted data after YYYY revert to default values. This newer ModDate should cue PDF viewers that the Info metadata is more current than the XMP data.

Alternatively, you might prefer to remove the XMP stream from the PDF altogether \(en that's what this option does.  Note that objects inside the PDF might have their own, separate XMP metadata streams, and that \fBdrop_xmp\fR does not remove those.  It only removes the PDF's document-level XMP stream.
.TP
.B [verbose]
By default, pdftk runs quietly. Append \fBverbose\fR to the end and it 
will speak up.
.TP
.B [dont_ask | do_ask]
Depending on the compile-time settings (see ASK_ABOUT_WARNINGS), pdftk might prompt you for
further input when it encounters a problem, such as a bad password. Override this default behavior
by adding \fBdont_ask\fR (so pdftk won't ask you what to do) or \fBdo_ask\fR (so pdftk will ask you what to do).

When running in \fBdont_ask\fR mode, pdftk will over-write files with its output without notice.
.SH EXAMPLES
.TP 2
.B Collate scanned pages
pdftk A=even.pdf B=odd.pdf shuffle A B output collated.pdf
.br
or if odd.pdf is in reverse order:
.br

CHANGELOG.md
## [3.3.0] - 2021-08-16

### Added
 - Support filling multi-valued list fields
 - Support replacing fonts in form fields

### Changed
 - Renamed namespace of internal iText library
 - Partially updated internal iText library code to 4.2.0
   - NB: might introduce regressions!

### Fixed
 - Crashes with non-conforming inputs

## [3.2.3] - 2021-06-16

### Added
 - Support excluding pages from a page range with ~
 - Support TextLeading parameter in form field appearances

### Changed
 - update_info adds a new Info dictionary if none is found

### Fixed
 - Crash with non-conforming encrypted inputs
 - Misleading URL in producer string
 - Better detection of xref table in non-conforming inputs
 - Better form detection (in particular forms created by pdftk cat)

## [3.2.2] - 2020-12-21

### Fixed
 - Build error with latest bcprov
 - Crash when flattening forms
 - Proper error message with missing passwords

### Changed
 - Better handling of long and non-ASCII form options
 - Report more annotation types

## [3.2.1] - 2020-10-26

### Fixed
 - Crash with incomplete files
 - Crash with invalid forms
 - Flatten form fields in order

## [3.2.0] - 2020-09-27

### Added
 - Allow specifying attachment metadata
 - Support AES encryption

### Changed
 - Attachments compatible with PDF/A-3u
 - Default encryption AES-128

### Fixed
 - Regression: allow reading background/stamp from stdin

## [3.1.3] - 2020-06-22

### Fixed
 - Crash when reading FDF files

## [3.1.2] - 2020-06-08

### Changed
 - Add a pattern if missing from burst output

### Fixed
 - Allow outputting pages read from stdin more than once
 - Do not warn if PageMediaDimensions is given in update_info
 - Crash when opening output files in burst fails

## [3.1.1] - 2020-05-01

### Fixed
 - Do not escape form fields if UTF-8 output selected

## [3.1.0] - 2020-04-26

### Added
 - update_info handles page metrics (media, rotation and labels)

### Fixed
 - Report page label indices correctly

## [3.0.10] - 2020-04-18

### Fixed
 - Crash in dump_data_annots
 - Load commons-lang-3 only if needed
 - Crash when copying malformed inputs

## [3.0.9] - 2020-01-13

### Added
 - Native image build option

### Fixed
 - Print an informative error if missing dependencies
 - Crash with newlines in arguments

## [3.0.8] - 2019-10-14

### Changed
 - Build for JRE version 1.7

## [3.0.7] - 2019-09-09

### Fixed
 - Crash involving passwords and file handles (java.lang.NullPointerException).

## [3.0.6] - 2019-06-04

### Fixed
 - Crash in burst with invalid file patterns.
 - Do not report some spurious form fields.
 - Escape more characters on XML-encoded reports.

## [3.0.5] - 2019-05-20

### Fixed
 - Do not print error if PdfID is missing.
 - Crash in burst with certain inputs (java.lang.NullPointerException).

## [3.0.4] - 2019-05-02

### Fixed
 - Crashes with type casting (java.lang.ClassCastException)

## [3.0.3] - 2019-01-15

### Fixed
 - Bug that corrupts images in PDF files with (de)compress option
 - Crash with incomplete records

## [3.0.2] - 2018-12-05

### Fixed
 - Issue with rotation not being applied.
 - Do not require owner password when user password is given.

## [3.0.1] - 2018-09-30

### Fixed
 - Issue reading one file three or more times.

## [3.0.0] - 2018-09-04

### Added
 - Translation of pdftk into Java.


java/com/gitlab/pdftk_java/pdftk.java
/*
 *   This file is part of the pdftk port to java
 *
 *   Copyright (c) Marc Vinyals 2017-2018
 *
 *   The program is a java port of PDFtk, the PDF Toolkit
 *   Copyright (c) 2003-2013 Steward and Lee, LLC
 *
 *   This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 2 of the License, or
 *   (at your option) any later version.
 *
 *   The program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.gitlab.pdftk_java;

import java.io.ByteArrayOutputStream;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.PrintStream;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.Scanner;

public class pdftk {

  /* TODO: should read from compiler */
  static final String PDFTK_VER = "3.3.0";
  static final boolean ASK_ABOUT_WARNINGS = false;

  static {
    System.setProperty("java.awt.headless", "true");
  }

  // For compatibility with Java < 9
  static void copyStream(InputStream inputStream, OutputStream outputStream) throws IOException {
    final int bufferSize = 0x2000;
    byte[] buffer = new byte[bufferSize];
    while (true) {
      int readBytes = inputStream.read(buffer, 0, bufferSize);
      if (readBytes < 0) break;
      outputStream.write(buffer, 0, readBytes);
    }
  }

  static byte[] readAllBytes(InputStream inputStream) throws IOException {
    ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
    copyStream(inputStream, outputStream);
    return outputStream.toByteArray();
  }

  static String prompt_for_password(String pass_name, String pass_app) {
    System.out.println("Please enter the " + pass_name + " password to use on " + pass_app + ".");
    System.out.println("   It can be empty, or have a maximum of 32 characters:");
    Scanner s = new Scanner(System.in);
    String password = s.nextLine();
    if (32 < password.length()) { // too long; trim
      System.out.println("The password you entered was over 32 characters long,");
      System.out.println("   so I am dropping: \"" + password.substring(32) + "\"");
      password = password.substring(0, 32);
    }
    return password;
  }

  static void warn_if_argument_is_password(String password) {
    boolean first = true;
    if (first && !password.isEmpty() && !password.equals("PROMPT")) {
      first = false;
      System.err.println(
          "Warning: Using a password on the command line interface can be insecure.");
      System.err.println("Use the keyword PROMPT to supply a password via standard input instead.");
    }
  }

  static String prompt_for_filename(String message) {
    // input could be multibyte, so try working
    // with bytes instead of formatted input features

    System.out.println(message);

    Scanner s = new Scanner(System.in);
    return s.nextLine();
  }

  static boolean confirm_overwrite(String filename) {
    System.out.println(
        "Warning: the output file: " + filename + " already exists.  Overwrite? (y/n)");
    Scanner s = new Scanner(System.in);
    String buff = s.nextLine();
    return buff.startsWith("y") || buff.startsWith("Y");
  }

  static boolean file_exists(String filename) {
    try {
      FileInputStream fp = new FileInputStream(filename);
      return true;
    } catch (FileNotFoundException e) {
      return false;
    }
  }

  static OutputStream get_output_stream(String output_filename, boolean ask_about_warnings_b) {
    if (output_filename.isEmpty() || output_filename.equals("PROMPT")) {
      output_filename = prompt_for_filename("Please enter a name for the output:");
      // recurse; try again
      return get_output_stream(output_filename, ask_about_warnings_b);
    }
    if (output_filename.equals("-")) { // stdout
      return System.out;
    }

    if (ask_about_warnings_b) {
      // test for existing file by this name
      boolean output_exists_b = false;
      if (file_exists(output_filename)) {
        if (!confirm_overwrite(output_filename)) {
          // recurse; try again
          return get_output_stream("PROMPT", ask_about_warnings_b);
        }
      }
    }

    return get_output_stream_file(output_filename);
  }

  static OutputStream get_output_stream_file(String output_filename) {
    OutputStream os_p = null;
    // attempt to open the stream
    try {
      os_p = new FileOutputStream(output_filename);
    } catch (IOException ioe_p) { // file open error
      System.err.println("Error: Failed to open output file: ");
      System.err.println("   " + output_filename);
      System.err.println("   No output created.");
      os_p = null;
    }
    return os_p;
  }

  static PrintStream get_print_stream(String output_filename, boolean output_utf8_b)
      throws IOException {
    Charset encoding = (output_utf8_b ? StandardCharsets.UTF_8 : StandardCharsets.US_ASCII);
    if (output_filename.isEmpty() || output_filename.equals("-")) {
      return new PrintStream(System.out, true, encoding.name());
    } else {
      return new PrintStream(output_filename, encoding.name());
    }
  }

  public static void main(String[] args) {
    System.exit(main_noexit(args));
  }

  public static int main_noexit(String[] args) {
    boolean help_b = false;
    boolean version_b = false;
    boolean synopsis_b = (args.length == 0);
    ErrorCode ret_val = ErrorCode.NO_ERROR; // default: no error

    for (String argv : args) {
      version_b = version_b || (argv.equals("--version")) || (argv.equals("-version"));
      help_b = help_b || (argv.equals("--help")) || (argv.equals("-help")) || (argv.equals("-h"));
    }

    if (help_b) {
      describe_full();
    } else if (version_b) {
      describe_header();
    } else if (synopsis_b) {
      describe_synopsis();
    } else {
      try {
        TK_Session session = new TK_Session();
        session.parse(args);

        session.dump_session_data();

        if (session.is_valid()) {
          // create_output() prints necessary error messages
          ret_val = session.create_output();
          if (ret_val == ErrorCode.ERROR) {
            System.err.println("   No output created.");
          } else if (ret_val == ErrorCode.PARTIAL) {
            System.err.println("   Not all outputs created.");
            ret_val = ErrorCode.ERROR;
          }
        } else { // error
          System.err.println("Done.  Input errors, so no output created.");
          ret_val = ErrorCode.ERROR;
        }
      }
      // per https://bugs.launchpad.net/ubuntu/+source/pdftk/+bug/544636
      catch (java.lang.ClassCastException c_p) {
        String message = c_p.getMessage();
        if (message == null) message = "";
        if (message.indexOf("com.lowagie.text.pdf.PdfDictionary") >= 0
            && message.indexOf("com.lowagie.text.pdf.PRIndirectReference") >= 0) {
          System.err.println("Error: One input PDF seems to not conform to the PDF standard.");
          System.err.println("Perhaps the document information dictionary is a direct object");
          System.err.println("   instead of an indirect reference.");
          System.err.println("Please report this bug to the program which produced the PDF.");
          System.err.println();
        }
        System.err.println("Java Exception:");
        c_p.printStackTrace();
        ret_val = ErrorCode.ERROR;
      } catch (NoClassDefFoundError error) {
        describe_missing_library(error);
        ret_val = ErrorCode.ERROR;
      } catch (java.lang.Throwable t_p) {
        System.err.println("Unhandled Java Exception in main():");
        t_p.printStackTrace();
        ret_val = ErrorCode.BUG;
      }
    }
    if (ret_val == ErrorCode.BUG) {
      describe_bug_report();
    }
    return ret_val.code;
  }

  static void describe_header() {
    System.out.println(
        "pdftk port to java " + PDFTK_VER + " a Handy Tool for Manipulating PDF Documents");
    System.out.println(
        "Copyright (c) 2017-2018 Marc Vinyals - https://gitlab.com/pdftk-java/pdftk");
    System.out.println("Copyright (c) 2003-2013 Steward and Lee, LLC.");
    System.out.println("pdftk includes a modified version of the iText library.");
    System.out.println("Copyright (c) 1999-2009 Bruno Lowagie, Paulo Soares, et al.");
    System.out.println(
        "This is free software; see the source code for copying conditions. There is");
    System.out.println(
        "NO warranty, not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.");
  }

  static void describe_resource(String resource) {
    InputStream inputStream = pdftk.class.getResourceAsStream(resource);
    try {
      copyStream(inputStream, System.out);
    } catch (IOException e) {
      e.printStackTrace();
      describe_bug_report();
    }
  }

  static void describe_synopsis() {
    describe_resource("resources/synopsis.txt");
  }

  static void describe_full() {
    describe_header();
    System.out.println();

    describe_synopsis();
    System.out.println();

    describe_resource("resources/description.txt");
  }

  static void describe_bug_report() {
    System.err.println("There was a problem with pdftk-java. Please report it at");
    System.err.println("https://gitlab.com/pdftk-java/pdftk/issues");
    System.err.println(
        "including the message above, the version of pdftk-java ("
            + PDFTK_VER
            + "), and if possible steps to reproduce the error.");
  }

  static void describe_missing_library(Throwable error) {
    System.err.println("Error: could not load a required library for this operation.");
    System.err.println(error);
    System.err.println("Make sure that bcprov and commons-lang3 are installed and included in the");
    System.err.println("classpath. See also https://gitlab.com/pdftk-java/pdftk/issues/2.");
  }
}
